src/navigation_controller.ts

```diff
@@ -139,29 +138,43 @@ export class NavigationController {
   }
 
   /**
-   * Sets whether the navigation controller should automatically intercept tab,
-   * enter, and escape navigation events to ensure that the user needs to
-   * explicitly enter and exit the Blockly environment for keyboard shortcuts to
-   * be enabled.
+   * Sets whether the navigation controller has focus. This has no side effect
+   * unless auto navigation is enabled via setHasAutoNavigationEnabled.
    *
-   * This mode has no effect unless setHasFocus is correctly used to communicate
-   * focus gained and lost contexts.
+   * @param isFocused whether the environment has browser focus.
+   */
+  setHasFocus(isFocused: boolean) {
+    this.hasNavigationFocus = isFocused;
+  }
+
+  /**
+   * Determines whether keyboard navigation should be allowed based on the
+   * current state of the workspace.
+   *
+   * A return value of 'true' generally indicates that the workspace both has
+   * enabled keyboard navigation and is currently in a state (e.g. focus) that
+   * can support keyboard navigation.
    *
-   * @param autoNavEnabled whether the controller should automatically
-   *     intercept navigation events.
+   * @param workspace the workspace in which keyboard navigation may be allowed.
+   * @returns whether keyboard navigation is currently allowed.
    */
-  setHasAutoNavigationEnabled(autoNavEnabled: boolean) {
-    this.isAutoNavigationEnabled = autoNavEnabled;
+  private isKeyboardNavigationCurrentlyAllowed(workspace: WorkspaceSvg) {
+    return workspace.keyboardAccessibilityMode && this.hasNavigationFocus;
   }
 
   /**
-   * Sets whether the navigation controller has focus. This has no side effect
-   * unless auto navigation is enabled via setHasAutoNavigationEnabled.
+   * Determines whether the provided workspace is currently keyboard navigable
+   * and editable.
    *
-   * @param isFocused whether the environment has browser focus.
+   * For the criteria on keyboard navigability, see
+   * isKeyboardNavigationCurrentlyAllowed.
+   *
+   * @param workspace the workspace in which keyboard editing may be allowed.
+   * @returns whether keyboard navigation and editing is currently allowed.
    */
-  setHasFocus(isFocused: boolean) {
-    this.hasNavigationFocus = isFocused;
+  private isWorkspaceCurrentlyKeyboardEditable(workspace: WorkspaceSvg) {
```

Comment:
I like the suggestion for 'can' since it's more deliberately tied to the input when read out. Will make this change in a follow-up.

I'm (unfortunately) using 'currently' to capture the entire context of whether the accessibility mode is currently enabled. One alternative is to use 'and' but I'm usually not a fan of that (since it feels like an obvious violation of single responsibility even though it's just a name--I try to find the common single behavior). I don't have a better idea than 'currently' at the moment. :) I lean toward ``canCurrentlyEdit``.